Frame tests for loading shared objects for the emscripten build

anutosh491 opened the pull request:
# Description

Now that https://github.com/llvm/llvm-project/pull/133037 has been merged, we should be able to test out loading shared libraries for the emscripten build.



## Type of change

Please tick all options which are relevant.

- [x] Bug fix
- [ ] New feature
- [ ] Requires documentation updates

## Testing

Please describe the test(s) that you added and ran to verify your changes.

## Checklist

- [x] I have read the contribution guide recently

mcbarton commented:
@anutosh491 Once this PR is ready will be able to run the DynamicLibraryManagerTest Sanity test? Currently its skipped.

anutosh491 commented on unittests/CppInterOp/DynamicLibraryManagerTest.cpp:
```diff
@@ -57,3 +57,25 @@ TEST(DynamicLibraryManagerTest, Sanity) {
+  EXPECT_NE(Addr, nullptr) << "Symbol 'ret_zero' not found after dlopen.";
+
+  using RetZeroFn = int (*)();
+  auto Fn = reinterpret_cast<RetZeroFn>(Addr);
```
I think this should be fine, the convertion from void* to a function pointer is justified here.

vgvassilev commented on lib/Interpreter/CppInterOpInterpreter.h:
```diff
@@ -412,6 +412,18 @@ class Interpreter {
+      llvm::errs() << "[cppinterop] Warning: 'lookup' has no effect on WASM.\n";
+    }
+    // In WASM: directly use Interpreter's LoadDynamicLibrary
+    if (auto Err = inner->LoadDynamicLibrary(filename.c_str())) {
```
I'd think we should set `lookup` to `false` and reuse the rest of the code. Why do we need this early return?

anutosh491 replied:
> reuse the rest of the code.

Hmm wait I'm confused. 

For wasm cases, we would just need to leverage the LoadDynamicLibrary support that was introduced here https://github.com/llvm/llvm-project/pull/133037

The DLM or the terminologies inside it won't work directly here (I see we go from `DLM->loadLibrary` to `lookupLibrary` where I see the sys call to start with so won't work anyways)

https://github.com/compiler-research/CppInterOp/blob/28ba16e97f58d0364b610bad9371bc4ea8e8b304/lib/Interpreter/DynamicLibraryManager.cpp#L292-L293

This fails with the following 
```
2: [ RUN      ] DynamicLibraryManagerTest.BasicSymbolLookup
2: Failure value returned from cantFail wrapped call
2: No such file or directory
2: UNREACHABLE executed at /Users/anutosh491/micromamba/envs/xeus-cpp-wasm-host/include/llvm/Support/Error.h:797!
2: Aborted()
```
Which is true as we're preloading the lib in the virtual file system as shown through [this comment](https://github.com/compiler-research/CppInterOp/pull/548/files#r2089416686) and this is looking for the lib locally (so unreachable)

anutosh491 commented on lib/Interpreter/CppInterOpInterpreter.h:
```diff
@@ -412,6 +412,18 @@ class Interpreter {
   }
 
   CompilationResult loadLibrary(const std::string& filename, bool lookup) {
+#ifdef __EMSCRIPTEN__
```
That's a thoughtful suggestion. Using target opts and checking for arch should be the way to go here (which I fully missed out on)

Thanks !

mcbarton replied:
If this is the route we plan to take here we should probably raise an issue to replace all the other occurances of  `ifdef __EMSCRIPTEN__` in the tests and code

anutosh491 replied:
Hmm, I think its still fine to do this for the tests (cause we're using emscripten to run the tests hahaha) but inside the source code, we should probably check the target triple as the loadLibrary implementation (or rather any functionality ) would be the arch/triple specific rather i.e wasm specific rather than based on emscripten as such.

But yeah feel free to look to open an issue if you think it would improve what we're trying to achieve overall !

mcbarton commented:
> ## [Codecov](https://app.codecov.io/gh/compiler-research/CppInterOp/pull/548?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=compiler-research) Report
> Attention: Patch coverage is `28.57143%` with `5 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 76.28%. Comparing base [(`28ba16e`)](https://app.codecov.io/gh/compiler-research/CppInterOp/commit/28ba16e97f58d0364b610bad9371bc4ea8e8b304?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=compiler-research) to head [(`54e0e14`)](https://app.codecov.io/gh/compiler-research/CppInterOp/commit/54e0e148aa7aae866a717474f8f94a9338d76502?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=compiler-research).
> 
> [Files with missing lines](https://app.codecov.io/gh/compiler-research/CppInterOp/pull/548?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=compiler-research)	Patch %	Lines
> [lib/Interpreter/CppInterOpInterpreter.h](https://app.codecov.io/gh/compiler-research/CppInterOp/pull/548?src=pr&el=tree&filepath=lib%2FInterpreter%2FCppInterOpInterpreter.h&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=compiler-research#diff-bGliL0ludGVycHJldGVyL0NwcEludGVyT3BJbnRlcnByZXRlci5o)	28.57%	[5 Missing ⚠️ ](https://app.codecov.io/gh/compiler-research/CppInterOp/pull/548?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=compiler-research)
> Additional details and impacted files
> 🚀 New features to boost your workflow:

The tests don't cover all the lines of the patch. Please fix this.

anutosh491 commented:
> The tests don't cover all the lines of the patch. Please fix this.

Weird

1) The support for loading shared libraries in llvm was added in llvm 20
2) So if you see the test caters only to 
```
TEST(DynamicLibraryManagerTest, BasicSymbolLookup) {
#ifndef EMSCRIPTEN
  GTEST_SKIP() << "This test is only intended for Emscripten builds.";
#else
  #if CLANG_VERSION_MAJOR < 20
    GTEST_SKIP() << "Support for loading shared libraries was added in LLVM 20.";
  #endif
#endif
```
3) So the added test on any `xx-xx-clang-rep-20-emscripten` would only pass if it goes through the mechanism added. 

I don't know how the above codecov report works but the test added is literally to address this :)

mcbarton commented:
> > The tests don't cover all the lines of the patch. Please fix this.
> 
> Weird
> 
> 1. The support for loading shared libraries in llvm was added in llvm 20
> 2. So if you see the test caters only to
> 
> ```
> TEST(DynamicLibraryManagerTest, BasicSymbolLookup) {
> #ifndef EMSCRIPTEN
>   GTEST_SKIP() << "This test is only intended for Emscripten builds.";
> #else
>   #if CLANG_VERSION_MAJOR < 20
>     GTEST_SKIP() << "Support for loading shared libraries was added in LLVM 20.";
>   #endif
> #endif
> ```
> 
> 3. So the added test on any `xx-xx-clang-rep-20-emscripten` would only pass if it goes through the mechanism added.
> 
> I don't know how the above codecov report works but the test added is literally to address this :)

Just had a look and this seems to be a result of us moving away from `ifdef __EMSCRIPTEN__` . Given the codecov is run on a native platform it won't ever be able test triple.isWasm() == true . @vgvassilev can decide what to do what that information is his review

![image](https://github.com/user-attachments/assets/cca3bcee-a91a-4c81-94da-f74019e969ca)

We probably need some way to have code coverage number for Emscripten if possible, but that is another issue to this PR entirely.

anutosh491 commented:
>  it won't ever be able test triple.isWasm() == true

Arghh, I thought using target opts here is a more organic solution but if we can't probably we can't :\ 
(should codecoverage be respected in every case ? if yes, can we some skip it for patches we know shouldn't be covered)

anutosh491 commented:
Hey @vgvassilev 

The other day I addressed your reviews here but we see this after I got rid of the hardcoding.

https://github.com/compiler-research/CppInterOp/pull/548#issuecomment-2881126442

To which I make this comment https://github.com/compiler-research/CppInterOp/pull/548#issuecomment-2881137380 saying what you suggested seems the organic way to get this but it might not be covered through codecov.

Let us know what you think !

vgvassilev reviewed (approved):
LGTM!

vgvassilev commented:
> > > The tests don't cover all the lines of the patch. Please fix this.
> > 
> > 
> > Weird
> > 
> > 1. The support for loading shared libraries in llvm was added in llvm 20
> > 2. So if you see the test caters only to
> > 
> > ```
> > TEST(DynamicLibraryManagerTest, BasicSymbolLookup) {
> > #ifndef EMSCRIPTEN
> >   GTEST_SKIP() << "This test is only intended for Emscripten builds.";
> > #else
> >   #if CLANG_VERSION_MAJOR < 20
> >     GTEST_SKIP() << "Support for loading shared libraries was added in LLVM 20.";
> >   #endif
> > #endif
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 3. So the added test on any `xx-xx-clang-rep-20-emscripten` would only pass if it goes through the mechanism added.
> > 
> > I don't know how the above codecov report works but the test added is literally to address this :)
> 
> Just had a look and this seems to be a result of us moving away from `ifdef __EMSCRIPTEN__` . Given the codecov is run on a native platform it won't ever be able test triple.isWasm() == true . @vgvassilev can decide what to do with that information is his review
> 
> ![image](https://private-user-images.githubusercontent.com/150042563/443790559-cca3bcee-a91a-4c81-94da-f74019e969ca.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc4MTA3MjgsIm5iZiI6MTc0NzgxMDQyOCwicGF0aCI6Ii8xNTAwNDI1NjMvNDQzNzkwNTU5LWNjYTNiY2VlLWE5MWEtNGM4MS05NGRhLWY3NDAxOWU5NjljYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDUyMVQwNjUzNDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMWUxNzE2ZmRmYzIwMTFmNWY2OTIwNTEwYmUzZTQzNTZmNjE3NDA5M2Y2NDg5NDI5MjQ3NjhhNWJiYmRlNDVhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Bg14QVM6niutkJdLiNN1l5IZy5XpCdL-KtV2KIN1QYg)
> 
> We probably need some way to have code coverage number for Emscripten if possible, but that is another issue to this PR entirely.

We should move away from that hardcoding but in a separate PR that covers the tests.
